master/buildbot/reporters/slack.py

```diff
@@ -0,0 +1,177 @@
+        #         'kinds': ('start', 'finish')
+        #     },
+        # }
+        self._mapping = mapping
+        # TODO: this goes to the documentation
+        # list of channels to post a message to if none are specified in the
+        # mapping.  the default behaviour is to use the channel configured for
+        # the webhook
+        self._default_channels = default_channels
+
+        self._http = yield HTTPClientService.getService(self.master, url)
+
+    def _filter_builds(self, build):
+        """
+        check if we should post a message for the given build:
+
+        * the builder is present in the map
+        * 'kinds' has the value for the kind of this notification (start/finish)
+        """
+        not_interested = None, None
+
+        props = self._mapping.get(build['builder'], dict())
+        if not props:
+            return not_interested
+
+        kind = 'finish' if build['complete'] else 'start'
+        if kind not in props['kinds']:
+            return not_interested
+
+        if self._default_channels is not _Default:
+            channels = props.get('channels', self._default_channels)
+            if not channels:
+                return not_interested
+        else:
+            channels = None
+
+        return props, channels
+
+    @staticmethod
+    def _prepare_attachment(build):
+        """
+        prepare an attachment based on the build information
+        """
+        if build['complete']:
+            result = build['results']
+
+            title = 'Build #{} has finished: {}'.format(build['buildid'],
+                                                        result)
+            # NOTE: we might want to include links to the logs of failed
+            # step(s) to 'text' => update neededDetails for the class
+            text = ' '.join(build.getText())
+            timestamp = build['complete_at']
+            color = 'good' if result == SUCCESS else 'danger'
+        else:
+            title = 'Build #{} has started'.format(build['buildid'])
+            text = None
+            timestamp = build['started_at']
+            color = 'warning'
+
+        attachment = dict(
+            title_name=title,
+            title_url=build['url'],
+            fallback='{}: <{}>'.format(title, build['url']),
+            color=color,
+            ts=timestamp
+        )
+
+        if text:
+            attachment['text'] = text
+
+        return attachment
+
+    @defer.inlineCallbacks
+    def send(self, build):
+        """
+        check the build and post a message to all configured channels
+        """
+        props, channels = self._filter_builds(build)
+        if not props:
+            defer.returnValue(SUCCESS)
+        else:
+            payload = dict(
+                username=self._username,
+                attachments=[self._prepare_attachment(build)])
+
+            posts = []
+
+            if channels is None:
+                posts.append(self._http.post("", json=payload))
```

Comment:
I would expect baseURL to be some URL to the base slack API, and here we would use the REST endpoint use to post the message.
Even if we use only ever one endpoint, I think it make sense to separate baseURL and endpoint.